### ISSUE TITLE
Feature Request for Nuxt 4 - Components Inside Pages Directory

### ISSUE THREAD

**Issue by imnash21** [nuxt/nuxt#28894]
### Describe the feature

Sometimes we need components inside pages directory.

Example:

```
- pages
  - login <- login page dir
    - _partials <- sub dir
      - form.vue <- sub component
    - index.vue <- login page

  - register <- register page dir
    - _partials <- sub dir
      - form.vue <- sub component
    - index.vue <- register page
```

What I mean is adding an auto import support inside pages dir only for directory with special character like underscore (_), double underscore (__) or any special characters that nuxt can determined that directory is not a page, so the auto import will scan that as component(s).

Then it will be like this:

```
<LoginForm />
<LazyLoginForm />

<RegisterForm />
<LazyRegisterForm />
```
The directory with underscore (_) will not be included to the component name.

### Additional information

- [ ] Would you be willing to help implement this feature?
- [ ] Could this feature be implemented as a module?

### Final checks

- [x] Read the [contribution guide](https://nuxt.com/docs/community/contribution).
- [X] Check existing [discussions](https://github.com/nuxt/nuxt/discussions) and [issues](https://github.com/nuxt/nuxt/issues).

**Comment by manniL** [comment]
Just as a heads up - Nuxt has [Layers](https://nuxt.com/docs/getting-started/layers) to have a better separation/slicing

**Comment by bbhxwl** [comment]
> Just as a heads up - Nuxt has [Layers](https://nuxt.com/docs/getting-started/layers) to have a better separation/slicing

Can't Layers only inherit templates for use? Can't we put the page and components in the same directory?

**Comment by danielroe** [comment]
This is possible - see https://deltener.com/blog/nuxt-enterprise-patterns-component-management/ for a starting point to implement this.

**Comment by imnash21** [comment]
> Just as a heads up - Nuxt has [Layers](https://nuxt.com/docs/getting-started/layers) to have a better separation/slicing

But the components inside the pages dir are very helpful to maintain a component that is specifically only to the page or as a part of the page only, it's easy to find and easy to remember where the component is.

**Comment by imnash21** [comment]
> This is possible with configuration - see https://deltener.com/blog/nuxt-enterprise-patterns-component-management/ for a starting point to implement this.

Nice, but still better if this configuration is already done, and ready to use.

**Comment by mathieunicolas** [comment]
This is interesting and I learned something thanks to this issue and the link provided by danielroe !

Since the configuration is not really painful, maybe it could simply consist in writing a markdown in the [Recipes](https://nuxt.com/docs/guide/recipes/) section of the doc ?
I mean, there are many ways of organizing files, should we rely on some kind of hardcoded presets, or learn how to define ours ?

**Comment by bbhxwl** [comment]
> Just as a heads up - Nuxt has [Layers](https://nuxt.com/docs/getting-started/layers) to have a better separation/slicing

I tested all the layers, aren't they independent? If you rewrite nuxt.config.ts in layers, it seems to overwrite the global configuration.

**Comment by DamianGlowala** [comment]
What works and scales for me really well (given the current state of things in Nuxt) is an approach to structure `components/` folder in a way that lets me predict given component's location upfront.

Your component might belong to different areas: layouts, pages, or be shared across multiple ones.

E.g. for a component `Foo.vue`:
- used in a layout located at `layouts/default.vue`, it would be located at `components/layouts/default/Foo.vue`
- used in a page located at `pages/posts/index.vue`, it would be located at `components/pages/posts/index/Foo.vue`
- used in a page located at `pages/posts/[id].vue`, it would be located at `components/pages/posts/[id]/Foo.vue`
- shared across multiple areas, it would be located at `components/shared/Foo.vue`

Example of components' places of usage:
```
- layouts/
  - default.vue
- pages/
  - posts/
    - index.vue
    - [id].vue
```

... and their corresponding locations:
```
- components/
  - layouts/
    - default/
      - Foo.vue
  - pages/
    - posts/
      - index/
        - Foo.vue
      - [id]/
        - Foo.vue
  - shared/
    - Foo.vue
```

**Comment by gulshan** [comment]
Personally, I am a big fan of organizing code/project by function concerns/features instead of technical concerns. This makes "getting into" a codebase really easy, at least for me. So, I love the workaround in the [linked blog](https://deltener.com/blog/nuxt-enterprise-patterns-component-management/).   
Still, I would like to see this as a built-in feature, perhaps a boolean in the nuxt config.

**Comment by GalacticHypernova** [comment]
The main issue I sense in this is that there are more than likely projects that use these patterns in actual pages. Unless a designated prefix that is invalid for a URL path would be used, this will be a breaking change. There's also a question of the obsoleteness of the components directory with such a change

**Comment by imnash21** [comment]
> The main issue I sense in this is that there are more than likely projects that use these patterns in actual pages. Unless a designated prefix that is invalid for a URL path would be used, this will be a breaking change. There's also a question of the obsoleteness of the components directory with such a change

**Sometimes** components inside the pages directory can help to organize specific components, the components specifically to the page as a partial, and the main/default components directory, u can use it for global/shared components like Button, Input, Card, Menu, Navbar, etc. so the default components directory will not be replaced.

**Comment by bbhxwl** [comment]
> > The main issue I sense in this is that there are more than likely projects that use these patterns in actual pages. Unless a designated prefix that is invalid for a URL path would be used, this will be a breaking change. There's also a question of the obsoleteness of the components directory with such a change
> 
> **Sometimes** components inside the pages directory can help to organize specific components, the components specifically to the page as a partial, and the main/default components directory, u can use it for global/shared components like Button, Input, Card, Menu, Navbar, etc. so the default components directory will not be replaced.

Yes, sometimes components are only used once, which can keep the page clean instead of writing all the components used once on the page. This is more suitable for developing large-scale projects. I feel that nuxt3 should have this method built-in.

**Comment by imnash21** [comment]
> > > The main issue I sense in this is that there are more than likely projects that use these patterns in actual pages. Unless a designated prefix that is invalid for a URL path would be used, this will be a breaking change. There's also a question of the obsoleteness of the components directory with such a change
> > 
> > 
> > **Sometimes** components inside the pages directory can help to organize specific components, the components specifically to the page as a partial, and the main/default components directory, u can use it for global/shared components like Button, Input, Card, Menu, Navbar, etc. so the default components directory will not be replaced.
> 
> Yes, sometimes components are only used once, which can keep the page clean instead of writing all the components used once on the page. This is more suitable for developing large-scale projects. I feel that nuxt3 should have this method built-in.

I'm waiting for this feature, hopefully this pattern will be supported (built-in) soon.

**Comment by gulshan** [comment]
As `Nuxt 4` can accommodate breaking changes, and as technically this feature can be considered breaking for the cases where chosen special name of the components directory being used as sub-path, this can be a built-in `Nuxt 4` feature. Although with flags, this can also be considered a non-breaking change.

**Comment by tux2nicolae** [comment]
Please considered this one.

We do have a pretty large project and unfortunately things get pretty ugly, we have a lot of components in the global components folder that are just used by some unpopular page, and as a side efect they pollutes the organisation and visibility for all truly global components.

This is somehow mandatory for the projects over a specific complexity threshold, the main reason people tend to use a framework like nuxt is probably the organisation and project structure, to predict with ease the design of an app and where the things are. In this particular case it seems more natural to grup pages and it's components together then by conceptual programming construct.

**Comment by arkhaiel** [comment]
> Please considered this one.
> 
> We do have a pretty large project and unfortunately things get pretty ugly, we have a lot of components in the global components folder that are just used by some unpopular page, and as a side efect they pollutes the organisation and visibility for all truly global components.
> 
> This is somehow mandatory for the projects over a specific complexity threshold, the main reason people tend to use a framework like nuxt is probably the organisation and project structure, to predict with ease the design of an app and where the things are. In this particular case it seems more natural to grup pages and it's components together then by conceptual programming construct.

Hi @tux2nicolae , I'm wondering, to elaborate some smart solution, why do you put your components in the global components folder ?

And I have another question, what is missing for your usecase when you follow the infos in [this page](https://deltener.com/blog/nuxt-enterprise-patterns-component-management/) ?

**Comment by tux2nicolae** [comment]
> Why do you put your components in the global components folder

I mean, we actually have a folder structure for the components but it can't be very big if we rely on components auto import because otherwise we'll have to use a name like this "VeryLongFolderStructureInsideTheComponentFolderMyComponentName"

This won't be that big of an issue because we don't really rely on autoimport for this kind of structures, more like for the really global components that are shared across multiple pages, but the biggest problem we see here is that when your navigating for some component you have to ignore a lot of irrelevant components that from our perspective should be grouped together with the page once it is used only for specific pages.

> And I have another question, what is missing for your usecase when you follow the infos in [this page](https://deltener.com/blog/nuxt-enterprise-patterns-component-management/) ?

@arkhaiel [this](https://deltener.com/blog/nuxt-enterprise-patterns-component-management/) would perfectly work for us but it seems more like a workaround than something standard supported by the framework

For now we'll try to use the above workaround, but it would be also nice to have something built-in directly in the framework

**Comment by HigherOrderLogic** [comment]
I'm actually against this.

You can't be sure that there'll be nobody would create routes names `/components`, `/_partials`. Also, I don't think it's easy to decide a work-for-all name for this folder.

Beside, setting up this structure is relatively easy and everyone can do it in their own project. If you want to simplify the process even more, it's possible to create a simple module.

**Comment by tux2nicolae** [comment]
@HigherOrderLogic yes, I get your point, indeed `/components` could be a valid page path. For my specific case probably I would use the above workaround.

I'm not really advocating for a specific implementation, just stated our scenario and that at some point it gets pretty overwhelmed. If it's decided to be implemented at some point there could be some other way, maybe reserve the `/(components)` path, or add a feature flag for `/components` path

**Comment by TheAlexLichter** [comment]
Hey all 👋🏻
Discussing this in the team, we have no intention to ship as default for Nuxt.
**But** we want to make it easier to configure for people preferring this setup.
As outlined in @hecktarzuli's [blog post](https://deltener.com/blog/nuxt-enterprise-patterns-component-management/) this is already possible in userland.

As changing the component auto imports "in pages" is already possible, the only thing that is tedious to set up is handling the "component pages". To make it easier (and also open up options for other use cases), it'd make sense to introduce a configuration option for the page scan pattern, so no hooks have to be called to remove "component pages".

**Comment by bbhxwl** [comment]
I want to temporarily define more convenient components on the webpage. Make the entire project more intuitive.

**Comment by bbhxwl** [comment]
> Hey all 👋🏻 Discussing this in the team, we have no intention to ship as default for Nuxt. **But** we want to make it easier to configure for people preferring this setup. As outlined in [@hecktarzuli](https://github.com/hecktarzuli)'s [blog post](https://deltener.com/blog/nuxt-enterprise-patterns-component-management/) this is already possible in userland.
> 
> As changing the component auto imports "in pages" is already possible, the only thing that is tedious to set up is handling the "component pages". To make it easier (and also open up options for other use cases), it'd make sense to introduce a configuration option for the page scan pattern, so no hooks have to be called to remove "component pages".

After writing in this way, the code prompts are not good, and it seems that Lazy's code prompts are not supported either. And he will generate a lot of imports

**Comment by gabrielstuff** [comment]
Very nice !

Discovered with https://www.youtube.com/watch?v=aAjfywmbfkM 

Putting the nuxt config just for helping others composing using components colocation.

```
export default defineNuxtConfig({
  compatibilityDate: '2025-05-01',
  devtools: { enabled: true },
  pages: {
    pattern: ['**/*.vue', '!**/components/**']
  },
  components: [
    '~/components', {
      path: '~/pages',
      pattern: '**/components/**',
      pathPrefix: false
    }
  ],
  app: {
    head: {
      htmlAttrs: {
        style: 'filter: invert(1)'
      }
    }
  }
})
```

**Comment by lkjimy** [comment]
> Very nice !
> 
> Discovered with https://www.youtube.com/watch?v=aAjfywmbfkM
> 
> Putting the nuxt config just for helping others composing using components colocation.
> 
> ```
> export default defineNuxtConfig({
>   compatibilityDate: '2025-05-01',
>   devtools: { enabled: true },
>   pages: {
>     pattern: ['**/*.vue', '!**/components/**']
>   },
>   components: [
>     '~/components', {
>       path: '~/pages',
>       pattern: '**/components/**',
>       pathPrefix: false
>     }
>   ],
>   app: {
>     head: {
>       htmlAttrs: {
>         style: 'filter: invert(1)'
>       }
>     }
>   }
> })
> ```

The issue with this solution is that we are still loading local (page related) components to the global scope. This works, but we will have to name components with huge name prefixes just to make sure the component names don't colide, specially in a big project.

Yes, we can remove the import in the `components` array (as in the example), but then, we lose HMR for all the content inside `/pages/my-page/components`.

In my setup, I'd like to keep HMR for the excluded directory, while manually importing it's content on the page.
Another solution is to simply add '-' as a prefix for the files, but it comes with the same issue: We lose HMR.
I was expecting the manual import to automatically be tracked, but that is not the case for excluded folders.

You may think we just have to reload the page (if we don't include the folder into auto imports), but in my experiments I had to restart the Nuxt server to see the changes I made. It's very unproductive, and much slower as the project grows.

If there is any workaround, please share.
